### PR TITLE
Handle spaces after request line

### LINKED ISSUE #41824: Kestrel does not allow spaces after the request line
A partner team reported compat issues with a client that is sending a space between the HTTP/1.1 request line and the first header. Kestrel (correctly) rejects this request as invalid. IIS allows this.

IIS allows:
```
GET / HTTP/1.1
 Host: example.com
User-Agent: curl
 ```
But not:
```
GET / HTTP/1.1
User-Agent: curl
 Host: example.com
```

They're working to get the client fixed, but it will take a while.

Proposal: Patch 6.0 to allow this under an app context switch but do not forward port the patch to main/7.0.

### PR DESCRIPTION
**Pull request by Tratcher**
# Handle spaces after request line

Handle malformed requests with a space between the request line and the first header

## Description

A partner team reported compat issues with a client that is sending a space between the HTTP/1.1 request line and the first header. Kestrel (correctly) rejects this request as invalid. IIS allows this. The client will eventually be fixed, but they've requested a 6.0 patch to unblock them for the next few months. We do not plan to port this change to main/7.0.

Fixes #41824

## Customer Impact

Preventing customer adoption of Kestrel due to client comapt issues.

## Regression?

- [ ] Yes
- [x] No

## Risk

- [ ] High
- [ ] Medium
- [x] Low

This is an opt in adjustment to the parser.

## Verification

- [x] Manual (required)
- [x] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [x] N/A

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs [hunk 1]
```diff
@@ -16,14 +16,27 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
...
+
+        internal HttpParser(bool showErrorDetails, bool allowSpaceAfterRequestLine)
         {
             _showErrorDetails = showErrorDetails;
+            _allowSpaceAfterRequestLine = allowSpaceAfterRequestLine;
+        }
+
+        private static bool CheckAllowSpaceAfterRequestLine()
+        {
+            AppContext.TryGetSwitch("Microsoft.AspNetCore.Server.Kestrel.AllowSpaceAfterRequestLine", out var allowSpaceAfterRequestLine);
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs:38]:
```suggestion
            // This workaround will be removed in .NET 7.
            AppContext.TryGetSwitch("Microsoft.AspNetCore.Server.Kestrel.AllowSpaceAfterRequestLine", out var allowSpaceAfterRequestLine);
```

Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs:38]:
```suggestion
            // This workaround will be removed in .NET 7.
            AppContext.TryGetSwitch("Microsoft.AspNetCore.Server.Kestrel.AllowSpaceAfterRequestLine", out var allowSpaceAfterRequestLine);
```

File: src/Servers/Kestrel/Core/test/HttpParserTests.cs [hunk 2]
```diff
@@ -52,6 +52,24 @@ public class HttpParserTests
...
         }
 
+        [Fact]
+        public void ParsesRequestLineWithTrailingSpace()
+        {
+            var parser = CreateParser(_nullTrace, allowSpaceAfterRequestLine: true);
+            var buffer = new ReadOnlySequence<byte>(Encoding.ASCII.GetBytes("GET / HTTP/1.1\r\n "));
+            var requestHandler = new RequestHandler();
+
+            Assert.True(ParseRequestLine(parser, requestHandler, buffer, out var consumed, out var examined));
```
Comment (halter73) [inline src/Servers/Kestrel/Core/test/HttpParserTests.cs:62]:
```suggestion
            Assert.False(ParseRequestLine(parser, requestHandler, buffer[..^1], out _, out _));
            Assert.True(ParseRequestLine(parser, requestHandler, buffer, out var consumed, out var examined));
```

Comment (halter73) [inline src/Servers/Kestrel/Core/test/HttpParserTests.cs:62]:
```suggestion
            Assert.False(ParseRequestLine(parser, requestHandler, buffer[..^1], out _, out _));
            Assert.True(ParseRequestLine(parser, requestHandler, buffer, out var consumed, out var examined));
```

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs [hunk 3]
```diff
@@ -48,7 +62,25 @@ public bool ParseRequestLine(TRequestHandler handler, ref SequenceReader<byte> r
 
             if (reader.TryReadTo(out ReadOnlySpan<byte> requestLine, ByteLF, advancePastDelimiter: true))
             {
+                if (_allowSpaceAfterRequestLine)
+                {
+                    // Skip a space after the request line
+                    if (reader.TryPeek(out byte s) && s == ByteSpace)
```
Comment (davidfowl) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs:68]: One space?

**Comment by Pilchie** [comment]
This was approved over email.

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs (hunk 3 above)
Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs:68]: Yes, from the reports we think that's all we need.

**Review by adityamandaleeka — APPROVED** [review]

**Pull request merged**